### ISSUE TITLE
getThemeInfo()--->HashtagCms\Models\Theme, null returned

### ISSUE THREAD

**Issue by ArcDee** [marghoobsuleman/hashtagcms#20]
Hello,

In the portal, I have done the following:
- Added one site (Site2) site_id=5
- Added one more language lang_id=5
- Added one more theme theme_id=5
- Configured the settings of the theme and categories
- Added one Blog and one Page in both languages

Thereafter, I got following error in BE/FE:

1.
TypeError
MarghoobSuleman\HashtagCms\Core\Main\InfoLoader::getThemeInfo(): Return value must be of type MarghoobSuleman\HashtagCms\Models\Theme, null returned
https://domain.com/login

2.
MarghoobSuleman\HashtagCms\Core\Main\InfoLoader::getThemeInfo
vendor/marghoobsuleman/hashtagcms/src/Core/Main/InfoLoader.php:403

3.
I observe the following:

In the table sites, I have two entries for Site1 & Site2 with the default language land_id=1 as well as theme_id=1 and theme_id=5. There is no data lang_id=5 related to Site2.

I created two entries by copy of both existing entries (as above) with identical data and changing the lang_id. 

Now there are two entries for lang_id=5 for both the sites. The theme_id and site_id is also there.

The above error continues. Any further ideas what I could do?

**Comment by ArcDee** [comment]
[stacktrace]
#0 /xxx/domain.com/public_html/hashtagcms/vendor/marghoobsuleman/hashtagcms/src/Core/Main/DataLoader.php(232): MarghoobSuleman\\HashtagCms\\Core\\Main\\InfoLoader->getThemeInfo()
#1 /xxx/domain.com/public_html/hashtagcms/vendor/marghoobsuleman/hashtagcms/src/Core/Main/LayoutManager.php(545): MarghoobSuleman\\HashtagCms\\Core\\Main\\DataLoader->loadData()
#2 /xxx/domain.com/public_html/hashtagcms/vendor/marghoobsuleman/hashtagcms/src/Core/Traits/FeEssential.php(24): MarghoobSuleman\\HashtagCms\\Core\\Main\\LayoutManager->getHTMLData()
#3 /xxx/domain.com/public_html/hashtagcms/vendor/marghoobsuleman/hashtagcms/src/Http/Controllers/LoginController.php(95): MarghoobSuleman\\HashtagCms\\Http\\Controllers\\FrontendBaseController->index()
#4 /xxx/domain.com/public_html/hashtagcms/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(36): MarghoobSuleman\\HashtagCms\\Http\\Controllers\\LoginController->index()
#5 /xxx/domain.com/public_html/hashtagcms/vendor/laravel/framework/src/Illuminate/Container/Util.php(40): Illuminate\\Container\\BoundMethod::Illuminate\\Container\\{closure}()
#6 /xxx/domain.com/public_html/hashtagcms/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(93): Illuminate\\Container\\Util::unwrapIfClosure()
#7 /xxx/domain.com/public_html/hashtagcms/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(37): Illuminate\\Container\\BoundMethod::callBoundMethod()
#8 /xxx/domain.com/public_html/hashtagcms/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(66): Illuminate\\Container\\BoundMethod::call()
#9 /xxx/domain.com/public_html/hashtagcms/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php(32): Illuminate\\Container\\BoundMethod::callClass()
#10 /xxx/domain.com/public_html/hashtagcms/vendor/laravel/framework/src/Illuminate/Container/Container.php(653): Illuminate\\Container\\BoundMethod::call()
#11 /xxx/domain.com/public_html/hashtagcms/vendor/marghoobsuleman/hashtagcms/src/routes/web.php(85): Illuminate\\Container\\Container->call()
#12 /xxx/domain.com/public_html/hashtagcms/vendor/laravel/framework/src/Illuminate/Routing/Route.php(238): Illuminate\\Support\\ServiceProvider->{closure}()
#13 /xxx/domain.com/public_html/hashtagcms/vendor/laravel/framework/src/Illuminate/Routing/Route.php(208): Illuminate\\Routing\\Route->runCallable()
#14 /xxx/domain.com/public_html/hashtagcms/vendor/laravel/framework/src/Illuminate/Routing/Router.php(695): Illuminate\\Routing\\Route->run()
#15 /xxx/domain.com/public_html/hashtagcms/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(128): Illuminate\\Routing\\Router->Illuminate\\Routing\\{closure}()
#16 /xxx/domain.com/public_html/hashtagcms/vendor/marghoobsuleman/hashtagcms/src/Core/Middleware/FeMiddleware.php(39): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#17 /xxx/domain.com/public_html/hashtagcms/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): MarghoobSuleman\\HashtagCms\\Core\\Middleware\\FeMiddleware->handle()
#18 /xxx/domain.com/public_html/hashtagcms/vendor/laravel/framework/src/Illuminate/Routing/Middleware/SubstituteBindings.php(50): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#19 /xxx/domain.com/public_html/hashtagcms/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Illuminate\\Routing\\Middleware\\SubstituteBindings->handle()
#20 /xxx/domain.com/public_html/hashtagcms/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/VerifyCsrfToken.php(78): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#21 /xxx/domain.com/public_html/hashtagcms/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Illuminate\\Foundation\\Http\\Middleware\\VerifyCsrfToken->handle()
#22 /xxx/domain.com/public_html/hashtagcms/vendor/laravel/framework/src/Illuminate/View/Middleware/ShareErrorsFromSession.php(49): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#23 /xxx/domain.com/public_html/hashtagcms/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Illuminate\\View\\Middleware\\ShareErrorsFromSession->handle()
#24 /xxx/domain.com/public_html/hashtagcms/vendor/laravel/framework/src/Illuminate/Session/Middleware/StartSession.php(121): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#25 /xxx/domain.com/public_html/hashtagcms/vendor/laravel/framework/src/Illuminate/Session/Middleware/StartSession.php(64): Illuminate\\Session\\Middleware\\StartSession->handleStatefulRequest()
#26 /xxx/domain.com/public_html/hashtagcms/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Illuminate\\Session\\Middleware\\StartSession->handle()
#27 /xxx/domain.com/public_html/hashtagcms/vendor/laravel/framework/src/Illuminate/Cookie/Middleware/AddQueuedCookiesToResponse.php(37): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#28 /xxx/domain.com/public_html/hashtagcms/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Illuminate\\Cookie\\Middleware\\AddQueuedCookiesToResponse->handle()
#29 /xxx/domain.com/public_html/hashtagcms/vendor/laravel/framework/src/Illuminate/Cookie/Middleware/EncryptCookies.php(67): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#30 /xxx/domain.com/public_html/hashtagcms/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Illuminate\\Cookie\\Middleware\\EncryptCookies->handle()
#31 /xxx/domain.com/public_html/hashtagcms/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(103): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#32 /xxx/domain.com/public_html/hashtagcms/vendor/laravel/framework/src/Illuminate/Routing/Router.php(697): Illuminate\\Pipeline\\Pipeline->then()
#33 /xxx/domain.com/public_html/hashtagcms/vendor/laravel/framework/src/Illuminate/Routing/Router.php(672): Illuminate\\Routing\\Router->runRouteWithinStack()
#34 /xxx/domain.com/public_html/hashtagcms/vendor/laravel/framework/src/Illuminate/Routing/Router.php(636): Illuminate\\Routing\\Router->runRoute()
#35 /xxx/domain.com/public_html/hashtagcms/vendor/laravel/framework/src/Illuminate/Routing/Router.php(625): Illuminate\\Routing\\Router->dispatchToRoute()
#36 /xxx/domain.com/public_html/hashtagcms/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(167): Illuminate\\Routing\\Router->dispatch()
#37 /xxx/domain.com/public_html/hashtagcms/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(128): Illuminate\\Foundation\\Http\\Kernel->Illuminate\\Foundation\\Http\\{closure}()
#38 /xxx/domain.com/public_html/hashtagcms/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TransformsRequest.php(21): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#39 /xxx/domain.com/public_html/hashtagcms/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/ConvertEmptyStringsToNull.php(31): Illuminate\\Foundation\\Http\\Middleware\\TransformsRequest->handle()
#40 /xxx/domain.com/public_html/hashtagcms/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Illuminate\\Foundation\\Http\\Middleware\\ConvertEmptyStringsToNull->handle()
#41 /xxx/domain.com/public_html/hashtagcms/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TransformsRequest.php(21): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#42 /xxx/domain.com/public_html/hashtagcms/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TrimStrings.php(40): Illuminate\\Foundation\\Http\\Middleware\\TransformsRequest->handle()
#43 /xxx/domain.com/public_html/hashtagcms/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Illuminate\\Foundation\\Http\\Middleware\\TrimStrings->handle()
#44 /xxx/domain.com/public_html/hashtagcms/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/ValidatePostSize.php(27): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#45 /xxx/domain.com/public_html/hashtagcms/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Illuminate\\Foundation\\Http\\Middleware\\ValidatePostSize->handle()
#46 /xxx/domain.com/public_html/hashtagcms/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/PreventRequestsDuringMaintenance.php(86): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#47 /xxx/domain.com/public_html/hashtagcms/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Illuminate\\Foundation\\Http\\Middleware\\PreventRequestsDuringMaintenance->handle()
#48 /xxx/domain.com/public_html/hashtagcms/vendor/fruitcake/laravel-cors/src/HandleCors.php(38): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#49 /xxx/domain.com/public_html/hashtagcms/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Fruitcake\\Cors\\HandleCors->handle()
#50 /xxx/domain.com/public_html/hashtagcms/vendor/laravel/framework/src/Illuminate/Http/Middleware/TrustProxies.php(39): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#51 /xxx/domain.com/public_html/hashtagcms/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Illuminate\\Http\\Middleware\\TrustProxies->handle()
#52 /xxx/domain.com/public_html/hashtagcms/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(103): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#53 /xxx/domain.com/public_html/hashtagcms/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(142): Illuminate\\Pipeline\\Pipeline->then()
#54 /xxx/domain.com/public_html/hashtagcms/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(111): Illuminate\\Foundation\\Http\\Kernel->sendRequestThroughRouter()
#55 /xxx/domain.com/public_html/hashtagcms/public/index.php(52): Illuminate\\Foundation\\Http\\Kernel->handle()
#56 {main}
"}

**Comment by ArcDee** [comment]
Environment information

Laravel version
    8.77.1
Laravel locale
    en
Laravel config cached
    false
PHP version
    8.0.14

**Comment by ArcDee** [comment]
Calling from the browser https://domain.com/blog, I get the error as follows:

popluated	"Error in calling controller (MarghoobSuleman\\HashtagCms\\Http\\Controllers\\BlogController@index)"
errorMessage	"Undefined array key \"category_link_rewrite\""

But in the table category, there are entries for both the sites on URI /blog with the default pattern {link_rewrite?}.

**Comment by ArcDee** [comment]
Log

Message
....Fetching category info : login

Level
....info

Query
....select
....  *
....from
....  `categories`
....where
....  (
........`link_rewrite` = ?
........and `publish_status` = ?
........and `site_id` = ?
....  )
....  and `site_id` = ?
....limit
....  1

Connection name
....mysql
0
....login
1
....1
2
....5
3
....1

Query
....select
....  *
....from
....  `categories`
....where
....  (
........`link_rewrite` = ?
........and `publish_status` = ?
........and `site_id` = ?
....  )
....  and `site_id` = ?
....limit
....  1

Connection name
....mysql
0
....login
1
....1
2
....5
3
....1

Log
Message
....login : Category not found!
Level
....info

**Comment by ArcDee** [comment]
I made a backup of the current database and executed the following:

php artisan cms:install

> Creating Tables...
> Seeding Tables...

So the database is reset to default. The Laravel installation remained unchanged.

The above mentioned error has not changed. If this would be an error coming from database, it should have gone. But this surprises me more because I have simply duplicated the basic them and renamed the directory of the theme. May be I should do something extra in there. Otherwise everything is default.

**Comment by ArcDee** [comment]
BTW, I have kept the old database as backup and seeded to a new database. 
Then, Laravel installation can work to the new database without the problems mentioned above.
I am able to work from scratch again.
So the problem is not in the scripts but in the database.

You can ask me for debugging purpose any questions to for e.g. run queries and find out something, if you want.

I now can recall that I had opened in the old database the setting of each site (two) and copied all default values available in the default to the new site. For e.g. I went through themes, modules, hooks, categories, etc. and made sure that they are available in the new site. In the new site, I could choose the theme, tenant, language, etc., thereafter. 

This means that these values were created in the new site.

Somehow, it is possible that I have missed something or some things were not created or something is missing. Here, you could ask me for running SQL queries for further debugging.

From the impression of this problem, I believe that it would be the best to provide cloning function that duplicates all values with one click after choosing a default site in the dropdown. Then it could clone all values instead of a user going through each tab and manually copy these values. Here, it should clone all areas of from the Master site to be cloned, including multiple languages and multiple categories in one click.

As these values are integral in the new site and can give a Kernel crash, it would be better to provide a sample site in multiple languages and multiple categories to clone for new site creation that does not crash the Kernel. This means the master (default) site can (and should) not be deleted.

One can expand the Master site languages and categories. But the basic (cannot be undeleted) scaffolding will remain. Thus, if I had created three languages and ten categories in the master site, I could easily clone into anathor ten new sites with one click.

I think this feature request - based on the nature of the above error - will become extremely powerful in the HashtagCMS.

**Comment by ArcDee** [comment]
Hello Merghoob,

I am one step further in trouble shooting. May be this will help you to narrow down the area to track.

I configured DomainOne.com and DomainTwo.com to resolve to the same public directory of HashtagCMS. Then, I followed following steps:

1. Create a new site DomainTwo.com
2. Edit the new site DomainTwo.com
3. Edit the new site DomainTwo.com
4. Created a new theme entry in DB.
5. Choose from drop down DomainTwo.com. Configure everywhere in categories (showing new IDs) and Pagemanager all necessary.
6. Confirm that all the settings in the default site DomainOne.com and DomainTwo.com is identical.

I did NOT create any new category or language in the new site DomainTwo.com, to have everything constant and track the error above.

The https://DomainTwo.com/login showed me a white page.

Thereafter, I copied the Hashtagcms directory and separated the php scripts to a new directory for the Domain.com, as follows:

DomainOne.com ---> /xxx/domainone.com/hashtagcms/public
DomainTwo.com ---> /xxx/domaintwo.com/hashtagcms/public

The database is the same. The scripts are the same. The configuration in .env, apps.php, etc is the same. Now, https://DomainTwo.com/login gives me following error:

```
    vendor/marghoobsuleman/hashtagcms/src/Core/Main/InfoLoader.php
    MarghoobSuleman\HashtagCms\Core\Main\InfoLoader:403

    vendor/marghoobsuleman/hashtagcms/src/Core/Main/DataLoader.php
    MarghoobSuleman\HashtagCms\Core\Main\DataLoader:232

    vendor/marghoobsuleman/hashtagcms/src/Core/Main/LayoutManager.php
    MarghoobSuleman\HashtagCms\Core\Main\LayoutManager:545

    vendor/marghoobsuleman/hashtagcms/src/Core/Traits/FeEssential.php
    MarghoobSuleman\HashtagCms\Http\Controllers\FrontendBaseController:24

    vendor/marghoobsuleman/hashtagcms/src/Http/Controllers/LoginController.php
    MarghoobSuleman\HashtagCms\Http\Controllers\LoginController

    Further stacktrace of Laravel scripts.

```
This error on DomainTwo.com is the same error I had earlier, while DomainOne.com is fully functional in BE and FE.

This shows that there is - most likely - a problem in the logic of HashtagCMS scripts. Some info is getting lost in querying something in the DB throwing this error.

**Comment by marghoobsuleman** [comment]
Okay, My bad, I've not shared the document for creating a new site. There are few steps to create a new site and everything. You don't have to do anything manually. Let me write and get back to you.

**Comment by ArcDee** [comment]
Hello Marghoob,

Thanks for the video on how to configure Multisite.  In my case, I had done the cloning and generating the records for a new site manually. I learned that it can be done by copy from old to new at the page level. This video will help many others.

I have seen your latest commits and did an update 1.3.0--->1.3.1. After that the Admin backend became accessible for the second site, which was not the case. But the frontend is showing an error.

From the view source, I found that the second domain cannot load css and js. Clicking on these links in view source, I could capture following error, which will help you to concentrate on the problem area:

/assets/hashtagcms/fe/dearnet/css/app.css
/assets/hashtagcms/fe/dearnet/js/app.js

{"popluated":"Error in calling controller (MarghoobSuleman\\HashtagCms\\Http\\Controllers\\FrontendController@index)",

"errorMessage":"Could not find the category.","errorTrace":[

{"file":"\/home\/xxx\/public_html\/hashtagcms\/vendor\/laravel\/framework\/src\/Illuminate\/Foundation\/helpers.php","line":44,"function":"abort","class":"Illuminate\\Foundation\\Application","type":"->"},

{"file":"\/home\/xxx\/public_html\/hashtagcms\/vendor\/marghoobsuleman\/hashtagcms\/src\/Core\/Traits\/FeEssential.php","line":26,"function":"abort"},

{"file":"\/home\/xxx\/public_html\/hashtagcms\/vendor\/marghoobsuleman\/hashtagcms\/src\/Http\/Controllers\/FrontendController.php","line":24,"function":"index","class":"MarghoobSuleman\\HashtagCms\\Http\\Controllers\\FrontendBaseController","type":"->"},

{"file":"\/home\/xxx\/public_html\/hashtagcms\/vendor\/laravel\/framework\/src\/Illuminate\/Container\/BoundMethod.php","line":36,"function":"index","class":"MarghoobSuleman\\HashtagCms\\Http\\Controllers\\FrontendController","type":"->"},

It seems that certain minor changes needs to be applied on the front end also.

Thanks for the update.

**Comment by ArcDee** [comment]
Hello Marghoob,

I have a good news!

After I installed HashtagCMS fresh in a new directory and database, the first site worked and second site did not. The error remained the same.

Now, after comparing the directories and databases of both, I found out that for some stupid reasons, two files were missing in the new installation (where it did not work), namely hashtagcms.php hashtagcmsadmin.php.

After manually copying both these files, the frontend and backend began to work IMMEDIATELY! This means that your fix did not get into the new install and that was the reason why it did not work.

After you have fixed this multisite issue in 1.3.1, now this issue should be closed.

**Comment by ArcDee** [comment]
I know why I did not get your multisite fix in the fresh installation. It had in composer:

"marghoobsuleman/hashtagcms": "^1.3"

This it installed stable branch tag 1.3.0.

Now, I have changed it to:

"marghoobsuleman/hashtagcms": "dev-master"

Now, it recognizes the development branch and got the fix: 6fd3984

There is too much to learn everywhere!

**Comment by ArcDee** [comment]
Now this issue should be closed. 

Multisite works like charm!

HashtagCMS is definitely one of the best CMS in the world, looking at the logic you have pumped into. Your work is just amazing, which this world has yet to discover...
